ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractLoader.scala

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.dao.events
+
+import akka.stream.{BoundedSourceQueue, Materializer, QueueOfferResult}
+import akka.stream.scaladsl.{Keep, Sink, Source}
+import com.daml.error.{ContextualizedErrorLogger, DamlContextualizedErrorLogger}
+import com.daml.error.definitions.LedgerApiErrors
+import com.daml.ledger.offset.Offset
+import com.daml.ledger.resources.ResourceOwner
+import com.daml.lf.value.Value.ContractId
+import com.daml.logging.LoggingContext
+import com.daml.metrics.api.MetricsContext
+import com.daml.metrics.{InstrumentedGraph, Metrics}
+import com.daml.platform.indexer.parallel.BatchN
+import com.daml.platform.store.backend.ContractStorageBackend
+import com.daml.platform.store.dao.DbDispatcher
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success}
+
+trait Loader[KEY, VALUE] {
+
+  def load(key: KEY)(implicit loggingContext: LoggingContext): Future[Option[VALUE]]
+
+}
+
+class AkkaStreamParallelBatchedLoader[KEY, VALUE](
+    batchLoad: Seq[(KEY, LoggingContext)] => Future[Map[KEY, VALUE]],
+    createQueue: () => Source[(KEY, LoggingContext, Promise[Option[VALUE]]), BoundedSourceQueue[
+      (KEY, LoggingContext, Promise[Option[VALUE]])
+    ]],
+    maxBatchSize: Int,
+    parallelism: Int,
+)(implicit executionContext: ExecutionContext, materializer: Materializer)
+    extends Loader[KEY, VALUE] {
+  private implicit val contextualizedErrorLogger: ContextualizedErrorLogger =
+    DamlContextualizedErrorLogger.forClass(this.getClass)
+
+  private val (queue, done) = createQueue() // TODO maybe set up autorestart?
+    .via(
+      BatchN(
+        maxBatchSize = maxBatchSize,
+        maxBatchCount = parallelism,
+      )
+    )
+    .mapAsyncUnordered(parallelism) { batch =>
+      batchLoad(
+        batch.view.map { case (key, loggingContext, _) => key -> loggingContext }.toSeq
+      ).transform {
+        case Success(resultMap) =>
+          batch.view.foreach { case (key, _, promise) =>
+            promise.success(resultMap.get(key))
+            ()
+          }
+          Success(())
+
+        case Failure(t) =>
+          batch.view.foreach { case (_, _, promise) =>
+            promise.failure(
+              LedgerApiErrors.InternalError
+                .Generic("Contract lookup failed", Some(t))
```

Comment:
wrapping the error from underneath, because the same Exception cannot be propagated to all gRPC recepients due to the known synchronization issue

Reply:
InternalError sounds harsh for a situation that can just be a transient DB error, no?

Reply:
oh yeah right, do you have maybe a suggestion how we can map the underlying error back?

Reply:
Not sure if we can easily do that. but:
* if the failure is a `StatusRuntimeException`, we could re-wrap it but we would need to care about copying everything also when the source error definition changes OR re-create the problematic part (the `Metadata is the one that was mutated in place)
* If it is a `RuntimeException` (undetected/unhandled error), we can let it behave like a LedgerApiErrors.InternalError, so either wrap it here or let it be wrapped in the `ErrorInterceptor` (I would do the former)

e.g. about copying the `Metadata`:
```
          t match {
            case s: StatusRuntimeException =>
              import scala.util.chaining._
              val newMetadata = new Metadata().tap(_.merge(s.getTrailers))
              new StatusRuntimeException(s.getStatus, newMetadata)
            case other => other
          }
```

Reply:
nice!!!

Reply:
took it as is